js/src/components/picc-meter.js

```diff
@@ -0,0 +1,193 @@
+
+        var bar = getBar(this);
+        var line = getLine(this);
+
+        if (typeof value !== 'number' || isNaN(value)) {
+          // console.log('bad value:', value);
+
+          // reset the bar
+          bar.style.setProperty('display', 'none');
+          bar.style.removeProperty('height');
+
+          // reset the line
+          line.style.setProperty('display', 'none');
+          line.style.removeProperty('bottom');
+
+          // classify and bail
+          classify(this, {
+            'no_data': true,
+            'above_average': false,
+            'below_average': false,
+            'about_average': false
+          });
+
+        } else {
+
+          classify(this, {no_data: false});
+
+          var scale = function(v) {
+            return (v - min) / (max - min);
+          };
+
+          var percent = function(v) {
+            return (scale(v) * 100).toFixed(1) + '%';
+          };
+
+          bar.style.removeProperty('display');
+          // prevent the bar from exceeding the height
+          bar.style.setProperty('height', percent(Math.min(value, max)));
+
+          var difference;
+          // console.log('average:', this.getAttribute('average'), '->', average);
+          if (isNaN(average)) {
+            line.style.setProperty('display', 'none');
+            classify(this, {
+              'above_average': false,
+              'below_average': false,
+              'about_average': false
+            });
+          } else {
+            difference = value - average;
+            line.style.removeProperty('display');
+            line.style.setProperty('bottom', percent(this.average));
+            var aboutThreshold = getAttr(this, 'about-threshold', .05) * (max - min);
+            classify(this, {
+              'above_average': difference > 0,
+              'below_average': difference < 0,
+              'about_average': Math.abs(difference) <= aboutThreshold
+            });
+          }
+
+        }
+
+        delete this.__timeout;
+
+        this.dispatchEvent(new CustomEvent('update'));
+      }},
+
+      min: {
+        get: function() {
+          return this.__min;
+        },
+        set: function(value) {
+          this.__min = number(value, 0);
+          deferUpdate(this);
+        }
+      },
+
+      max: {
+        get: function() {
+          return this.__max;
+        },
+        set: function(value) {
+          this.__max = number(value, 1);
+          deferUpdate(this);
+        }
+      },
+
+      value: {
+        get: function() {
+          return this.__value;
+        },
+        set: function(value) {
+          this.__value = number(value);
+          deferUpdate(this);
+        }
+      },
+
+      average: {
+        get: function() {
+          return this.__average;
+        },
+        set: function(value) {
+          this.__average = number(value);
+          deferUpdate(this);
+        }
+      }
+    }
+  )
+});
+
+function deferUpdate(meter, delay) {
```

Comment:
Many of these functions are defined elsewhere. Can they be moved to a common utility module?